utils/warnings.ml

```diff
@@ -201,252 +201,328 @@ type description =
     (* The first element of the list is the current name, any following ones are
        deprecated. The current name should always be derived mechanically from
        the constructor name. *)
-    description : string; }
+    description : string;
+    since : string option
```

Comment:
Since 4.14, the module `Sys` defines a, `ocaml_release_info` type for ocaml version. It sounds like it would be a good fit here.

Reply:
Ok, I tried using it, but this adds a fair amount of code for the pretty-printing of the `ocaml_release_info` type; I didn't find any pretty-printers, so I tried writing my own:

```ocaml
let print_since_ocaml_release ppf since_opt =
  match since_opt with
  | None -> Printf.ifprintf ppf ""
  | Some { Sys.major; minor; patchlevel; extra } ->
      let print_extra ppf = function
        | None -> Printf.ifprintf ppf ""
        | Some (prefix, extra_str) ->
            let prefix_str = match prefix with
              | Sys.Plus -> "+"
              | Sys.Tilde -> "~"
            in
            Printf.fprintf ppf "%s%s" prefix_str extra_str
      in
      Printf.fprintf ppf " (since %d.%02d%s%a)"
        major
        minor
        (if patchlevel <> 0 then "." ^ string_of_int patchlevel else "")
        print_extra extra
```

There are a few assumptions here:
- I assume that the patchlevel is not useful unless non-zero, so I don't print it by default;
- To keep the old "4.0X" tradition, I pad the `minor` field with a zero. However, I believe that from version 5 onwards, this is no longer the case; so the pretty-printer should probably take both cases in account: `if major > 4 ... else ...`.

Also, each warning with a non-None `since` now looks like:

    since = Some {major = 4; minor = 0; patchlevel = 0; extra = None } };
    
Instead of `since = Some "4.00"`.

Would it be useful to provide a initializer such as `let make_ocaml_release ?patchlevel ?extra major minor`, so that the entries would be reduced to `since = make_ocaml_release 4 0` ?

It feels like such functions should have their place in the `Sys` module, to be more widely reusable.

Reply:
First, since you only want to print the major and minor version, it would be enough to define a somplified version in `warnings.ml`:
```ocaml
let pp_since out release_info = 
 Printf.printf "(since %d.%d) " release_info.major release_info.minor
```
This version can be extended to support the right padding version with
```ocaml
let pp_since out release_info =
  Printf.printf out "(since %d.%0*d) "
    release_info.major
    (if release_info.major >= 5 then 0 else 2)
    release_info.minor
```
Similarly, we can specialize  a `make_ocaml_release` to fit the invariant in `Warnings`:
```ocaml
let since major minor = { Sys.major; minor; patchlevel=0; extra=None }
```
It might make sense to have more generic function for `release_info` in `Sys`.
But we can start small with the exact function needed in `Warnings`, and reimplement them with the generic stdlib function later on if they end up being available.

Reply:
Thanks for the suggestions, I incorporated them. I used `Format` instead of `Printf`, however, since I needed `pp_print_option`, which is not available in the `Printf` module.

I considered adding an entry to the `Changes `file, but since the initial discussion was on Discord and not an issue in this Github, I'm not sure if it should be added. Or maybe it's sufficiently "not-quite-visible" to the user that it does not require an entry?

By the way, I added a second commit, fixing 2 missing dots in warning descriptions, to ensure they are all homogeneous.